### PR TITLE
Support for HTTP-Proxy

### PR DESCRIPTION
**Pull request by ccpp**
My Pull request adds functionality to use FreeRDP behind an HTTP proxy.
Enable the CMake option `WITH_HTTP_PROXY` to build.
There is a command line option `/http-proxy:<host>:<port>`.
For Connections using a gateway, also the `$https_proxy` / `%HTTPS_PROXY%` environment variables are respected.
Tested to work with squid and with TSG.

### CODE REVIEW

**Comment by ccpp** [comment]
Hello!

After some time the feature is brought up to date and bugfixed; the CMake option has been dropped as suggested.

Now, the buffered BIO is used to communicate with the proxy (only writing; For reading the reply, BIO_gets might be nice but it is not implemented in the buffered BIO).

In tcp.c, the connection to the proxy is now logically separated from all other connection code which keeps changing a lot.

And there are no more removed LoC :)
Can you test the build, please? Thanks, Chris.

**Comment by MatthiasLohr** [comment]
Any news about that topic? I would really like this feature!

**Comment by MatthiasLohr** [comment]
@ccpp, would it be possible for you to implement support for proxy authentication (username + password in proxy settings, like ```http://user:password@192.168.1.1:8080/)?

**Comment by ccpp** [comment]
Hi Matthias,

a lot has happened in FreeRDP since the branch was written and was prefectly mergeable. In the meantime, some conflicts must be resolved. Due to my lack of time (and little hope of getting accepted this time) I won't do this in nearest future.
Also, my last commit 91eddce introduced an in-source-build dependency which is bad...

You might want to implement proxy authentication on top of commit 3b7b3a0 :)

-Chris

**Comment by ccpp** [comment]
Hi again,
the pull request is up to date again. please test :)
(sorry, no proxy authentication so far, but I would see this as a separate feature / pull-request)

**Comment by akallabeth** [comment]
@freerdp-bot test

File: libfreerdp/core/proxy.c [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+	return FALSE;
+}
+
+void proxy_read_environment(rdpSettings *settings, char *envname)
+{
+	char env[256];
+	DWORD envlen;
+
+	envlen = GetEnvironmentVariableA(envname, env, sizeof(env)-1);
```
Comment (akallabeth) [inline libfreerdp/core/proxy.c]: Maybe use an allocated buffer to avoid introducing arbitrary limitations?

**Comment by bmiklautz** [comment]
@freerdp-bot test.

**Comment by ccpp** [comment]
...So there seems to be an issue wit `strndup` on windows. Which function is used there?
(How about `BUF_strdup` / `BUF_memdup` from OpenSSL - I thes this should work)

**Comment by mfleisz** [comment]
@ccpp You are right - strndup is not supported on Windows platforms. I guess the best way (which will work on all platforms) is to malloc and memcpy/strncpy.

**Comment by ccpp** [comment]
@mfleisz @akallabeth Both issues were fixed. Can you test it again?

**Comment by MatthiasLohr** [comment]
@ccpp Thank you very much! :)

I'm not a great C developer, so i'm sorry that i can't help to improve FreeRDP...

**Comment by akallabeth** [comment]
@ccpp One final request, you're using a different coding style in your new files. Could you run ```scripts/format_code.sh``` (requires astyle >= 2.04) on them?
@freerdp-bot would you be kind enough to actually complete a test?

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@bmiklautz Could you have a look why this is not running the tests?

**Comment by bmiklautz** [comment]
@freerdp-bot test.

**Comment by akallabeth** [comment]
@hardening @bmiklautz Anyone want to check/merge this one?

**Comment by bmiklautz** [comment]
@ccpp sorry it took soooo long. Thanks for your effort!

**Comment by ccpp** [comment]
Thanks for merging :)

**Pull request merged**